src/test/java/com/wikia/webdriver/common/logging/Log.java

```diff
@@ -0,0 +1,338 @@
+  }
+
+  public static void startTest(Method testMethod) {
+    String testName = testMethod.getName();
+    String className = testMethod.getDeclaringClass().getCanonicalName();
+    String command;
+    String description;
+    if (testMethod.isAnnotationPresent(RelatedIssue.class)) {
+      String issueID = testMethod.getAnnotation(RelatedIssue.class).issueID();
+      String jiraPath = "https://wikia-inc.atlassian.net/browse/";
+      String jiraUrl = jiraPath + issueID;
+      String jiraLink = VelocityWrapper.fillLink(jiraUrl, issueID);
+      command = "Known failure";
+      description =
+          testName + " - " + jiraLink + " " + testMethod.getAnnotation(RelatedIssue.class)
+              .comment();
+    } else {
+      command = "";
+      description = testName;
+    }
+    String html = VelocityWrapper.fillFirstLogRow(className, testName, command, description);
+    CommonUtils.appendTextToFile(LOG_PATH, html);
+    testStarted = true;
+  }
+
+  public static void logAssertionStacktrace(AssertionError exception) {
+    WikiaWebDriver driver = DriverProvider.getActiveDriver();
+
+    Log.imageCounter += 1;
+    if ("true".equals(Configuration.getLogEnabled())) {
+      String exceptionMessage = ExceptionUtils.getStackTrace(exception);
+      List<LogData> classList = new ArrayList<>();
+      classList.add(LogLevel.ERROR);
+      classList.add(LogType.STACKTRACE);
+      String html = VelocityWrapper.fillErrorLogRow(classList, exceptionMessage, Log.imageCounter);
+      try {
+        new Shooter().savePageScreenshot(Log.SCREEN_PATH + Log.imageCounter, driver);
+        CommonUtils.appendTextToFile(Log.SCREEN_PATH + Log.imageCounter + ".html",
+                                     Log.getPageSource(driver));
+        CommonUtils.appendTextToFile(Log.LOG_PATH, html);
+      } catch (Exception e) {
+        html = VelocityWrapper.fillErrorLogRowWoScreenshotAndSource(classList, exceptionMessage);
+        CommonUtils.appendTextToFile(Log.LOG_PATH, html);
+        Log.log("onException",
+                "driver has no ability to catch screenshot or html source - driver may died<br/>",
+                false);
+      }
+      Log.logJSError();
+    }
+  }
+
+  public static void stacktrace(Throwable throwable) {
+    String exceptionMessage = Throwables.getStackTrace(throwable);
+    List<LogData> classList = new ArrayList<>();
+    classList.add(LogLevel.ERROR);
+    classList.add(LogType.STACKTRACE);
+
+    VelocityWrapper
+        .fillLogRow(classList, "STACKTRACE", escapeHtml(exceptionMessage).replace("\n", "<br>"));
+  }
+
+  public static void stop() {
+    WikiaWebDriver driver = DriverProvider.getActiveDriver();
+    if (driver.getProxy() != null && Configuration.getForceHttps()) {
+      Har har = driver.getProxy().getHar();
+      for (HarEntry entry : har.getLog().getEntries()) {
+        URL url;
+        try {
+          url = new URL(entry.getRequest().getUrl());
+          if (url.getHost().contains("wikia")) {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar-s4/coding_rules#rule_key=squid%3AS134)